### PR TITLE
[TD-206] Automated smoke tests

### PR DESCRIPTION
**Pull request by abhishiekgowtham**
Adds a GitHub Action which runs smoke tests for PRs and master branch. Also, any suggestions on how we could test the hybrid chart?

### CODE REVIEW

File: .github/workflows/smoke-test.yml
Comment (abhishiekgowtham) [inline .github/workflows/smoke-test.yml:122]: Need someone with higher access to add this secret for the runner.

File: .github/kind.yml
Comment (excieve) [inline .github/kind.yml:13]: Is there a reason to have more than 1 CP node for tests?
Also, can it be a matrix of versions, so that we could cover different supported k8s versions (e.g. latest 1.16-1.20)?

File: .github/workflows/smoke-test.yml
Comment (excieve) [inline .github/workflows/smoke-test.yml:110]: `rpcKey` and `apiKey` should be secrets. It will need a permanent Tyk Cloud deployment it could connect too though — please raise an internal ticket (`TD`) to set one up.

**Review by excieve — COMMENTED** [review]
Thanks @abhishiekgowtham!

Hybrid can be done, but needs a permanent Tyk Cloud deployment for this. Alternatively, a Pro chart with MDCB enabled can be deployed and a hybrid test configured for that chart's MDCB endpoint if possible. That's slightly different than Cloud hybrid, but essentially the same thing would be tested.

Also, shouldn't we enable everything possible (portal, MDCB, etc.) to test all the things?

File: .github/kind.yml
Comment (abhishiekgowtham) [reply .github/kind.yml:13]: I wanted to replicate the testing as much close to production as possible so that's why i chose to run in HA for CP.

Comment (gernest) [reply .github/kind.yml:13]: Wont one `control-plane` and  two `worker` node satisfy minimum production  setting?

Comment (abhishiekgowtham) [reply .github/kind.yml:13]: i don't think that's HA or really production level. But running with 5 node setup there seems to be a [bug](https://github.com/kubernetes-sigs/kind/issues/1288) in KinD so will use a 3 node setup.

**Comment by abhishiekgowtham** [comment]
@excieve I have added support to run on multiple kubernetes versions. And various components like mdcb, tib, etc... are enabled during the upgrade not on install. 

I haven't enabled ingresses though.

File: .github/workflows/smoke-test.yml
Comment (gwotg) [inline .github/workflows/smoke-test.yml:108]: I'm running https://charts.bitnami.com/bitnami in a single node replicaset internally (not like in my mind). It both allows expansion to a real production mongo deployment and compatibility with our application (transaction). Should offshore the job of maintaining a deployment.

File: .github/workflows/smoke-test.yml
Comment (gwotg) [reply .github/workflows/smoke-test.yml:115]: Unblock yourself - split between what you can and cannot do without a permanent deployment, release the former and hang the later on the  ticket.

**Review by gwotg — CHANGES_REQUESTED** [review]
I think the split up - with / without a permanent deployment - is needed before you carry on with merging - I'd even say they are distinct and failing the later shouldn't block, as the availability of that deployment isn't quite off hours pager material.

File: .github/workflows/smoke-test.yml
Comment (gwotg) [reply .github/workflows/smoke-test.yml:108]: Relevant bits from my values

`
clusterDomain: cluster.local
architecture: replicaset
useStatefulSet: true

auth:
  enabled: true
  rootPassword: "<thou shalt not>"
  replicaSetKey: ""
replicaSetName: rs0
replicaSetHostnames: true
replicaCount: 1

persistence:
  enabled: false

service:
  nameOverride: ""
  type: ClusterIP
  port: 27017
  portName: mongodb
  nodePort: ""
  externalIPs: []
  loadBalancerSourceRanges: []
  annotations: {}

externalAccess:
  enabled: true
  service:
    type: LoadBalancer
    port: 27017
    loadBalancerIPs: 
      - {{ metallb_ip_ini }}
    domain: ara.dev

arbiter:
  enabled: true
`

Comment (gwotg) [reply .github/workflows/smoke-test.yml:108]: Yeah, I dunno how to format yaml in a comment.

Comment (abhishiekgowtham) [reply .github/workflows/smoke-test.yml:108]: i don't think we need the bitnami charts as it uses StatefulSets + a bunch of other stuff. A simple deployments should suffice. However, we have redis/mongodb charts in the cloudsmith that could be used.

File: .github/workflows/smoke-test.yml
Comment (abhishiekgowtham) [reply .github/workflows/smoke-test.yml:115]: The entire hybrid chart can't be tested without permanent deployment. I mean can set dummy values like i did, but that's not the point of having smoke-tests.

I believe it's better to knock it off in a single ticket rather than coming back to it in sometime with a different ticket.

**Review by excieve — APPROVED** [review]
Looks good! Feel free to merge once hybrid link is available.

**Comment by abhishiekgowtham** [comment]
@excieve Thanks for the approval. Merging this PR now.

**Comment by joshblakeley** [comment]
@abhishiekgowtham Can we please document how we test the chart in the README?

**Comment by abhishiekgowtham** [comment]
@joshblakeley Automated testing is meant for contributors so its kind of unnecessary to have it in the README. It should be part of the CONTRIBUTING.md. 

@TykTechnologies/devops what do you think?

**Pull request merged**
